File: android/src/main/java/com/sensorworks/RNAttitude/RNAttitudeModule.java
// Useful info:
// http://plaw.info/articles/sensorfusion/
// https://dev.opera.com/articles/w3c-device-orientation-usage/
// http://bediyap.com/programming/convert-quaternion-to-euler-rotations/
// https://stackoverflow.com/questions/19239482/using-quaternion-instead-of-roll-pitch-and-yaw-to-track-device-motion
// https://stackoverflow.com/questions/11103683/euler-angle-to-quaternion-then-quaternion-to-euler-angle
// https://stackoverflow.com/questions/14482518/how-multiplybyinverseofattitude-cmattitude-class-is-implemented
// https://stackoverflow.com/questions/5782658/extracting-yaw-from-a-quaternion
package com.sensorworks.RNAttitude;

import android.content.Context;
import android.os.SystemClock;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorManager;
import android.hardware.SensorEventListener;
import android.util.Log;

import java.util.Arrays;

import com.facebook.react.bridge.ActivityEventListener;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.BaseActivityEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Callback;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.modules.core.RCTNativeAppEventEmitter;

@ReactModule(name = RNAttitudeModule.NAME)
public class RNAttitudeModule extends ReactContextBaseJavaModule implements LifecycleEventListener,
    SensorEventListener {
  public static final String NAME = "RNAttitude";
  private static final double NS2MS = 0.000001;

  private static final byte ROTATE_NONE = 0;
  private static final byte ROTATE_LEFT = 1;
  private static final byte ROTATE_RIGHT = 2;

  private final ReactApplicationContext reactContext;
  private final Sensor rotationSensor;
  private final SensorManager sensorManager;
  private int intervalMillis;
  private long nextSampleTime;
  private long rotation;
  private boolean isRunning;
  private float pitchOffset;
  private float rollOffset;
  private float[] eulerAngles = new float[2];

  public RNAttitudeModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
    this.reactContext.addLifecycleEventListener(this);
    sensorManager = (SensorManager) reactContext.getSystemService(Context.SENSOR_SERVICE);
    rotationSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
    isRunning = false;
    intervalMillis = 200;
    nextSampleTime = 0;
    rotation = ROTATE_NONE;
    pitchOffset = 0;
    rollOffset = 0;
  }

  @Override
  public String getName() {
    return NAME;
  }

  @Override
  public void onAccuracyChanged(Sensor sensor, int accuracy) {}

  @Override
  public void onHostResume() {
    if(isRunning) {
      sensorManager.registerListener(this, rotationSensor, intervalMillis * 1000);
    }
  }

  @Override
  public void onHostPause() {
    if(isRunning) {
      sensorManager.unregisterListener(this);
    }
  }

  @Override
  public void onHostDestroy() {
    stopObserving();
  }

  //------------------------------------------------------------------------------------------------
  // React interface

  @ReactMethod
  // Determines if this device is capable of providing attitude updates
  public void isSupported(Promise promise) {
    promise.resolve(rotationSensor != null);
  }

  @ReactMethod
  // Applies an inverse to core data and compensates for any non-zero installations.
  // Basically it makes a new 'zero' position for both pitch and roll.
  public void zero() {
    pitchOffset = -eulerAngles[0];
    rollOffset = -eulerAngles[1];
  }

  @ReactMethod
  // Resets the pitch and roll offsets
  public void reset() {
    pitchOffset = 0;
    rollOffset = 0;
  }

  @ReactMethod
  // Sets the interval between event samples
  public void setInterval(int interval) {
    intervalMillis = interval;
    boolean shouldStart = isRunning;
    stopObserving();
    if(shouldStart) {
      startObserving(null);
    }
  }

  @ReactMethod
  // Sets the device rotation to either 'none', 'left' or 'right'
  // If this isn't called then we assume 'portrait'/no rotation orientation
  public void setRotation(String rotation) {
    String lowercaseRotation = rotation.toLowerCase();
    switch (lowercaseRotation) {
      case "none":
        this.rotation = ROTATE_NONE;
        break;
      case "left":
        this.rotation = ROTATE_LEFT;
        break;
      case "right":
        this.rotation = ROTATE_RIGHT;
        break;
      default:
        Log.e("ERROR", "Unrecognised rotation passed to react-native-attitude, must be 'none','left' or 'right' only");
        break;
    }
    reset();
  }

  @ReactMethod
  public void startObserving(Promise promise) {
    if (rotationSensor == null) {
      promise.reject("-1", "Rotation vector sensor not available; will not provide orientation data.");
      return;
    }
    nextSampleTime = 0;
    sensorManager.registerListener(this, rotationSensor, intervalMillis * 1000);
    isRunning = true;
    if(promise != null) {
      promise.resolve(intervalMillis);
    }
  }

  @ReactMethod
  public void stopObserving() {
    sensorManager.unregisterListener(this);
    isRunning = false;
  }

  //------------------------------------------------------------------------------------------------
  // Internal methods

  @Override
  public void onSensorChanged(SensorEvent sensorEvent) {
    float[] rotationMatrix = new float[9];
    float[] remappedMatrix = new float[9];
    float[] orientation = new float[3];

    // Time to run?
    long currentTime = SystemClock.elapsedRealtime();
    if (currentTime < nextSampleTime) {
      return;
    }

    // Get the current attitude value as a rotation matrix
    SensorManager.getRotationMatrixFromVector(rotationMatrix, getVectorFromSensorEvent(sensorEvent));

    // Remap the coordinate system depending on screen orientation
    if (rotation == ROTATE_LEFT) {
      SensorManager.remapCoordinateSystem(rotationMatrix, SensorManager.AXIS_Z, SensorManager.AXIS_MINUS_X, remappedMatrix);
    } else if (rotation == ROTATE_RIGHT) {
      SensorManager.remapCoordinateSystem(rotationMatrix, SensorManager.AXIS_MINUS_Z, SensorManager.AXIS_X, remappedMatrix);
    } else {
      SensorManager.remapCoordinateSystem(rotationMatrix, SensorManager.AXIS_X, SensorManager.AXIS_Z, remappedMatrix);
    }

    float heading = (float) (((((Math.toDegrees(SensorManager.getOrientation(remappedMatrix, orientation)[0]) + 360) % 360) -
        (Math.toDegrees(SensorManager.getOrientation(remappedMatrix, orientation)[2]))) + 360) % 360);

    // apply any pitch and roll offsets
    if(pitchOffset != 0 || rollOffset != 0) {
      float[] offsetMatrix = applyPitchOffset(pitchOffset, remappedMatrix);
      offsetMatrix = applyRollOffset(rollOffset, offsetMatrix);
      eulerAngles = getOrientation(offsetMatrix);
    }
    else {
      eulerAngles = getOrientation(remappedMatrix);
    }

    WritableMap map = Arguments.createMap();
    map.putDouble("timestamp", sensorEvent.timestamp * NS2MS);
    map.putDouble("roll", eulerAngles[1]);
    map.putDouble("pitch", eulerAngles[0]);
    map.putDouble("heading", heading);
    try {
      reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("attitudeUpdate", map);
    } catch (RuntimeException e) {
      Log.e("ERROR", "Error sending event over the React bridge");
    }

    // Calculate the next time we should run
    nextSampleTime = currentTime + intervalMillis;
  }

  private float[] getVectorFromSensorEvent(SensorEvent event) {
    if (event.values.length > 4) {
      // On some Samsung devices SensorManager.getRotationMatrixFromVector
      // appears to throw an exception if rotation vector has length > 4.
      // For the purposes of this class the first 4 values of the
      // rotation vector are sufficient (see crbug.com/335298 for details).
      // Only affects Android 4.3
      return Arrays.copyOf(event.values, 4);
    } else {
      return event.values;
    }
  }

  // Rotates the supplied rotation matrix so it is expressed in a different coordinate system.
  private float[] remapCoordinateSystem(float[] inR, int X, int Y) {
    float[] outR = new float[9];
    // Z is "the other" axis, its sign is either +/- sign(X)*sign(Y)
    // this can be calculated by exclusive-or'ing X and Y; except for
    // the sign inversion (+/-) which is calculated below.
    int Z = X ^ Y;
    // extract the axis (remove the sign), offset in the range 0 to 2.
    int x = (X & 0x3) - 1;
    int y = (Y & 0x3) - 1;
    int z = (Z & 0x3) - 1;
    // compute the sign of Z (whether it needs to be inverted)
    int axis_y = (z + 1) % 3;
    int axis_z = (z + 2) % 3;
    if (((x ^ axis_y) | (y ^ axis_z)) != 0) {
      Z ^= 0x80;
    }
    boolean sx = (X >= 0x80);
    boolean sy = (Y >= 0x80);
    boolean sz = (Z >= 0x80);
    // Perform R * r, in avoiding actual muls and adds.
    for (int j = 0; j < 3; j++) {
      int offset = j * 3;
      for (int i = 0; i < 3; i++) {
        if (x == i) outR[offset + i] = sx ? -inR[offset + 0] : inR[offset + 0];
        if (y == i) outR[offset + i] = sy ? -inR[offset + 1] : inR[offset + 1];
        if (z == i) outR[offset + i] = sz ? -inR[offset + 2] : inR[offset + 2];
      }
    }
    return outR;
  }

  // Computes the device's orientation based on the rotation matrix.
  // R should be double[9] array representing a rotation matrix
  private float[] getOrientation(float[] R) {
    // /  R[ 0]   R[ 1]   R[ 2]  \
    // |  R[ 3]   R[ 4]   R[ 5]  |
    // \  R[ 6]   R[ 7]   R[ 8]  /
    float[] out = new float[2];
    out[0] = (float) Math.toDegrees(Math.asin(R[7])); // pitch
    out[1] = (float) Math.toDegrees(Math.atan2(-R[6], R[8])); // roll
    return out;
  }

  // Apply a rotation about the roll axis to this rotation matrix.
  // see http://planning.cs.uiuc.edu/node102.html
  private float[] applyRollOffset(float roll, float[] matrixIn) {
    float value = (float) Math.toRadians(roll);
    float[] rotateMatrix = {
        (float) Math.cos(value), 0, (float) Math.sin(value),
        0, 1, 0,
        (float) -Math.sin(value), 0, (float) Math.cos(value)
    };
    return matrixMultiply(matrixIn, rotateMatrix);
  }

  // Apply a rotation about the pitch axis to this rotation matrix.
  // see http://planning.cs.uiuc.edu/node102.html
  private float[] applyPitchOffset(float pitch, float[] matrixIn) {
    float value = (float) Math.toRadians(pitch);
    float[] rotateMatrix = {
        1, 0, 0,
        0, (float) Math.cos(value), (float) -Math.sin(value),
        0, (float) Math.sin(value), (float) Math.cos(value)
    };
    return matrixMultiply(matrixIn, rotateMatrix);
  }

  // multiplies two rotation matrix, A and B
  private float[] matrixMultiply(float[] A, float[] B) {
    float[] result = new float[9];;
    result[0] = A[0] * B[0] + A[1] * B[3] + A[2] * B[6];
    result[1] = A[0] * B[1] + A[1] * B[4] + A[2] * B[7];
    result[2] = A[0] * B[2] + A[1] * B[5] + A[2] * B[8];
    result[3] = A[3] * B[0] + A[4] * B[3] + A[5] * B[6];
    result[4] = A[3] * B[1] + A[4] * B[4] + A[5] * B[7];
    result[5] = A[3] * B[2] + A[4] * B[5] + A[5] * B[8];
    result[6] = A[6] * B[0] + A[7] * B[3] + A[8] * B[6];
    result[7] = A[6] * B[1] + A[7] * B[4] + A[8] * B[7];
    result[8] = A[6] * B[2] + A[7] * B[5] + A[8] * B[8];
    return result;
  }

}

File: android/src/main/java/com/sensorworks/rnattitude/RNAttitudeModule.kt
package com.sensorworks.rnattitude

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.SystemClock
import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.WritableMap
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.round
import kotlin.math.sin

class RNAttitudeModule(reactContext: ReactApplicationContext) :
  NativeRNAttitudeSpec(reactContext), LifecycleEventListener, SensorEventListener {

  private val sensorManager: SensorManager =
    reactContext.getSystemService(Context.SENSOR_SERVICE) as SensorManager
  private val rotationSensor: Sensor? =
    sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)

  private var intervalMillis = 200
  private var nextSampleTime = 0L
  private var lastEmitTimeMs = 0L
  private var rotation = ROTATE_NONE
  private var output = OUTPUT_BOTH
  private var isRunning = false
  private var pitchOffset = 0f
  private var rollOffset = 0f
  private var headingLast = 0f
  private val eulerAngles = FloatArray(2)
  private val eulerAnglesLast = FloatArray(2)
  private val rotationMatrix = FloatArray(9)
  private val remappedMatrix = FloatArray(9)
  private val orientation = FloatArray(3)
  private val pitchAdjustedMatrix = FloatArray(9)
  private val rollAdjustedMatrix = FloatArray(9)

  init {
    reactContext.addLifecycleEventListener(this)
  }

  override fun getName(): String = NAME

  override fun isSupported(promise: Promise) {
    promise.resolve(rotationSensor != null)
  }

  override fun getAvailableSensors(promise: Promise) {
    val types =
      listOf(
        Sensor.TYPE_ACCELEROMETER to "accelerometer",
        Sensor.TYPE_GYROSCOPE to "gyroscope",
        Sensor.TYPE_MAGNETIC_FIELD to "magnetometer",
        Sensor.TYPE_ROTATION_VECTOR to "rotationVector",
      )
    val result = Arguments.createArray()
    for ((type, id) in types) {
      for (sensor in sensorManager.getSensorList(type)) {
        val map = Arguments.createMap()
        map.putString("id", id)
        map.putString("name", sensor.name)
        map.putString("vendor", sensor.vendor)
        map.putDouble("version", sensor.version.toDouble())
        map.putDouble("maxRange", sensor.maximumRange.toDouble())
        map.putDouble("resolution", sensor.resolution.toDouble())
        map.putDouble("minDelayUs", sensor.minDelay.toDouble())
        result.pushMap(map)
      }
    }
    promise.resolve(result)
  }

  override fun zero() {
    pitchOffset = -eulerAngles[0]
    rollOffset = -eulerAngles[1]
  }

  override fun reset() {
    pitchOffset = 0f
    rollOffset = 0f
  }

  override fun setOutput(outputIn: String) {
    val shouldStart = isRunning
    stopObserving()
    output =
      when (outputIn.lowercase()) {
        "both" -> OUTPUT_BOTH
        "attitude" -> OUTPUT_ATTITUDE
        "heading" -> OUTPUT_HEADING
        else -> {
          Log.e(
            TAG,
            "Unrecognised output passed to react-native-attitude, must be 'both', 'attitude' or 'heading' only"
          )
          output
        }
      }
    if (shouldStart) {
      startObserving()
    }
  }

  override fun setInterval(interval: Double) {
    intervalMillis = intervalMillisFor(interval)
    nextSampleTime = 0L
    lastEmitTimeMs = 0L
    val shouldStart = isRunning
    stopObserving()
    if (shouldStart) {
      startObserving()
    }
  }

  override fun setRotation(rotationIn: String) {
    rotation =
      when (rotationIn.lowercase()) {
        "none" -> ROTATE_NONE
        "left" -> ROTATE_LEFT
        "right" -> ROTATE_RIGHT
        else -> {
          Log.e(
            TAG,
            "Unrecognised rotation passed to react-native-attitude, must be 'none','left' or 'right' only"
          )
          rotation
        }
      }
    reset()
  }

  override fun startObserving() {
    if (rotationSensor == null) {
      return
    }
    nextSampleTime = 0L
    lastEmitTimeMs = 0L
    val samplingUs = samplingPeriodUs()
    sensorManager.registerListener(this, rotationSensor, samplingUs, samplingUs)
    isRunning = true
  }

  override fun stopObserving() {
    sensorManager.unregisterListener(this)
    isRunning = false
    eulerAngles[0] = 0f
    eulerAngles[1] = 0f
    eulerAnglesLast[0] = Float.NaN
    eulerAnglesLast[1] = Float.NaN
    headingLast = Float.NaN
    lastEmitTimeMs = 0L
  }

  override fun addListener(eventName: String) {}

  override fun removeListeners(count: Double) {}

  override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}

  override fun onHostResume() {
    if (isRunning && rotationSensor != null) {
      val samplingUs = samplingPeriodUs()
      sensorManager.registerListener(this, rotationSensor, samplingUs, samplingUs)
    }
  }

  override fun onHostPause() {
    if (isRunning) {
      sensorManager.unregisterListener(this)
    }
  }

  override fun onHostDestroy() {
    stopObserving()
  }

  override fun onSensorChanged(event: SensorEvent) {
    val currentTime = SystemClock.elapsedRealtime()
    if (currentTime < nextSampleTime) {
      return
    }

    SensorManager.getRotationMatrixFromVector(
      rotationMatrix,
      getVectorFromSensorEvent(event)
    )

    when (rotation) {
      ROTATE_LEFT ->
        SensorManager.remapCoordinateSystem(
          rotationMatrix,
          SensorManager.AXIS_Z,
          SensorManager.AXIS_MINUS_X,
          remappedMatrix
        )
      ROTATE_RIGHT ->
        SensorManager.remapCoordinateSystem(
          rotationMatrix,
          SensorManager.AXIS_MINUS_Z,
          SensorManager.AXIS_X,
          remappedMatrix
        )
      else ->
        SensorManager.remapCoordinateSystem(
          rotationMatrix,
          SensorManager.AXIS_X,
          SensorManager.AXIS_Z,
          remappedMatrix
        )
    }

    var heading = 0f

    if (output == OUTPUT_BOTH || output == OUTPUT_ATTITUDE) {
      val computed =
        if (pitchOffset != 0f || rollOffset != 0f) {
          applyPitchOffset(pitchOffset, remappedMatrix, pitchAdjustedMatrix)
          applyRollOffset(rollOffset, pitchAdjustedMatrix, rollAdjustedMatrix)
          getOrientation(rollAdjustedMatrix)
        } else {
          getOrientation(remappedMatrix)
        }
      eulerAngles[0] = round(computed[0] * 10) / 10f
      eulerAngles[1] = round(computed[1] * 10) / 10f
    } else {
      eulerAngles[0] = 0f
      eulerAngles[1] = 0f
    }

    if (output == OUTPUT_BOTH || output == OUTPUT_HEADING) {
      val azimuth = SensorManager.getOrientation(remappedMatrix, orientation)[0]
      heading = round(((Math.toDegrees(azimuth.toDouble()) + 360) % 360)).toFloat()
    }

    val nowMs = System.currentTimeMillis()
    val changed =
      eulerAnglesLast[0].isNaN() ||
        eulerAngles[0] != eulerAnglesLast[0] ||
        eulerAngles[1] != eulerAnglesLast[1] ||
        heading != headingLast
    val heartbeatDue =
      !changed &&
        lastEmitTimeMs > 0 &&
        nowMs - lastEmitTimeMs >= HEARTBEAT_INTERVAL_MS

    if (changed || heartbeatDue) {
      val map: WritableMap =
        Arguments.createMap().apply {
          putDouble("timestamp", nowMs.toDouble())
          putDouble("roll", eulerAngles[1].toDouble())
          putDouble("pitch", eulerAngles[0].toDouble())
          putDouble("heading", heading.toDouble())
        }
      emitOnAttitudeUpdate(map)
      lastEmitTimeMs = nowMs
      if (changed) {
        eulerAnglesLast[0] = eulerAngles[0]
        eulerAnglesLast[1] = eulerAngles[1]
        headingLast = heading
      }
    }

    nextSampleTime = currentTime + intervalMillis
  }

  private fun intervalMillisFor(interval: Double): Int =
    when (interval.toInt()) {
      1000, 200, 100, 50, 25 -> interval.toInt()
      else -> 200
    }

  private fun samplingPeriodUs(): Int {
    val sensor = rotationSensor ?: return intervalMillis * 1000
    val requestedUs = intervalMillis * 1000
    val minDelayUs = sensor.minDelay.coerceAtLeast(1000)
    return maxOf(requestedUs, minDelayUs)
  }

  private fun getVectorFromSensorEvent(event: SensorEvent): FloatArray {
    return if (event.values.size > 4) {
      event.values.copyOf(4)
    } else {
      event.values
    }
  }

  private fun getOrientation(matrix: FloatArray): FloatArray {
    val pitch = Math.toDegrees(asin(matrix[7].coerceIn(-1f, 1f).toDouble())).toFloat()
    val roll =
      Math.toDegrees(atan2(-matrix[6].toDouble(), matrix[8].toDouble())).toFloat()
    return floatArrayOf(pitch, roll)
  }

  private fun applyRollOffset(
    roll: Float,
    matrixIn: FloatArray,
    matrixOut: FloatArray
  ) {
    val value = Math.toRadians(roll.toDouble()).toFloat()
    val rotateMatrix =
      floatArrayOf(
        cos(value),
        0f,
        sin(value),
        0f,
        1f,
        0f,
        -sin(value),
        0f,
        cos(value)
      )
    matrixMultiply(matrixIn, rotateMatrix, matrixOut)
  }

  private fun applyPitchOffset(
    pitch: Float,
    matrixIn: FloatArray,
    matrixOut: FloatArray
  ) {
    val value = Math.toRadians(pitch.toDouble()).toFloat()
    val rotateMatrix =
      floatArrayOf(
        1f,
        0f,
        0f,
        0f,
        cos(value),
        -sin(value),
        0f,
        sin(value),
        cos(value)
      )
    matrixMultiply(matrixIn, rotateMatrix, matrixOut)
  }

  private fun matrixMultiply(
    a: FloatArray,
    b: FloatArray,
    result: FloatArray
  ) {
    result[0] = a[0] * b[0] + a[1] * b[3] + a[2] * b[6]
    result[1] = a[0] * b[1] + a[1] * b[4] + a[2] * b[7]
    result[2] = a[0] * b[2] + a[1] * b[5] + a[2] * b[8]
    result[3] = a[3] * b[0] + a[4] * b[3] + a[5] * b[6]
    result[4] = a[3] * b[1] + a[4] * b[4] + a[5] * b[7]
    result[5] = a[3] * b[2] + a[4] * b[5] + a[5] * b[8]
    result[6] = a[6] * b[0] + a[7] * b[3] + a[8] * b[6]
    result[7] = a[6] * b[1] + a[7] * b[4] + a[8] * b[7]
    result[8] = a[6] * b[2] + a[7] * b[5] + a[8] * b[8]
  }

  companion object {
    const val NAME = NativeRNAttitudeSpec.NAME
    private const val TAG = "RNAttitude"
    private const val HEARTBEAT_INTERVAL_MS = 1000L
    private const val ROTATE_NONE = 0
    private const val ROTATE_LEFT = 1
    private const val ROTATE_RIGHT = 2
    private const val OUTPUT_BOTH = 0
    private const val OUTPUT_ATTITUDE = 1
    private const val OUTPUT_HEADING = 2
  }
}
